### PR TITLE
Added an isinstance check to fix limits not working with symbols which have zero = True

### LINKED ISSUE #13715: Limit can fail if an expression contains a zero=True symbol
>>> from sympy import *
    >>> n = Symbol('n')
    >>> p = Symbol('p', zero=True)
    >>> limit(n+p, n, 0)
    Traceback (most recent call last):
      File "<pyshell#3>", line 1, in <module>
        limit(n+p, n, 0)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\series\limits.py", line 49, in limit
        return Limit(e, z, z0, dir).doit(deep=False)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\series\limits.py", line 180, in doit
        r = gruntz(e, z, z0, dir)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\series\gruntz.py", line 658, in gruntz
        r = limitinf(e0, z)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\core\cache.py", line 93, in wrapper
        retval = cfunc(*args, **kwargs)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\series\gruntz.py", line 428, in limitinf
        c0, e0 = mrv_leadterm(e, x)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\core\cache.py", line 93, in wrapper
        retval = cfunc(*args, **kwargs)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\series\gruntz.py", line 513, in mrv_leadterm
        series = calculate_series(f, w, logx=logw)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\series\gruntz.py", line 466, in calculate_series
        for t in e.lseries(x, logx=logx):
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\core\expr.py", line 2645, in yield_lseries
        o = Order(si, x)*x
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\core\cache.py", line 93, in wrapper
        retval = cfunc(*args, **kwargs)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\series\order.py", line 213, in __new__
        lst = expr.extract_leading_order(args)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\core\cache.py", line 93, in wrapper
        retval = cfunc(*args, **kwargs)
      File "C:\Users\E\AppData\Local\Programs\Python\Python36\lib\site-packages\sympy\core\add.py", line 783, in extract_leading_order
        if o.contains(of) and o != of:
    AttributeError: 'Zero' object has no attribute 'contains'

### PR DESCRIPTION
**Pull request by NikhilPappu**
#### References to other Issues or PRs

This PR Fixes #13715.

#### Brief description of what is fixed or changed
Added an isinstance check to fix limits not working with symbols which have zero = True. It seems a Zero object is being passed instead of an Order object in that case.

#### Other comments

* core
  * Limits are now working with symbols which have `zero = True`.

### CODE REVIEW

**Comment by NikhilPappu** [comment]
Is this an acceptable fix or does it come at the cost of too much extra computation?

File: sympy/core/add.py [hunk 1]
```diff
@@ -781,9 +781,10 @@ def extract_leading_order(self, symbols, point=None):
         seq = [(f, Order(f, *zip(symbols, point))) for f in self.args]
         for ef, of in seq:
             for e, o in lst:
-                if o.contains(of) and o != of:
-                    of = None
-                    break
+                if(isinstance(o, Order)):
```
Comment (asmeurer) [inline sympy/core/add.py]: Don't use parentheses with `if`.

**Comment by asmeurer** [comment]
Doing a type check seems fine, but the question is if the code is still correct with this change.

**Comment by asmeurer** [comment]
We also should investigate if Order is actually doing the right thing, or if the fix should be made there instead. 

Some more explanation of the source of the problem and why the fix is correct would be nice here.

**Comment by NikhilPappu** [comment]
A Zero object seems to be passed down instead of an Order object when the expression contains a symbol with zero = True. I can't seem to figure out why. The check added gives this output
`p = symbols('p', zero = True)
    assert limit(x + p, x, 2)  == 2`
Should the expected output be p or 0?

**Review by arighnac — COMMENTED** [review]
Remove the trailing whitespace from line 104 in test_limits.py

File: sympy/core/add.py [hunk 2]
```diff
@@ -781,9 +782,10 @@ def extract_leading_order(self, symbols, point=None):
         seq = [(f, Order(f, *zip(symbols, point))) for f in self.args]
         for ef, of in seq:
             for e, o in lst:
-                if o.contains(of) and o != of:
-                    of = None
-                    break
+                if not isinstance(o, Zero):
```
Comment (gxyd) [inline sympy/core/add.py]: should this be `if not o.is_zero:`?

File: sympy/core/add.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 from .cache import cacheit
 from .numbers import ilcm, igcd
 from .expr import Expr
+from .numbers import Zero
```
Comment (jksuom) [inline sympy/core/add.py]: Is Zero used?

Comment (NikhilPappu) [reply sympy/core/add.py]: It isn't. I'll remove it.

**Comment by czgdp1807** [comment]
@asmeurer What is the status of this PR?

**Comment by czgdp1807** [comment]
ping @asmeurer https://github.com/sympy/sympy/pull/13735#issuecomment-533774582

**Comment by czgdp1807** [comment]
ping @jksuom Your views on [this](https://github.com/sympy/sympy/pull/13735#issuecomment-533774582).

**Comment by jksuom** [comment]
I'm not sure what `.extract_leading_order` is expected to do. The docstring says:
> Returns the leading term and its order.

However, it may return several terms if they are of the same order.
```>>> a = Symbol('a')
>>> (2*x + a*x).extract_leading_order(x)
((a*x, O(x)), (2*x, O(x)))
```
Moreover, possible cancellations are not considered:
```
>>> (x - sin(x) + x**2).extract_leading_order(x)
((-sin(x), O(x)), (x, O(x)))
```
There is no indication that the order of the expression is `O(x**2)` (with leading term `x**2`). This will lead to the wrong result
```
>>> Order(x - sin(x) + x**2, x)
O(2*x)  # from O(x) + O(x)
```
here in order.py:
https://github.com/sympy/sympy/blob/4fc50841e38205561523e0a9d0e5da173988b04a/sympy/series/order.py#L219-L221

So, there are issues in the extraction of the leading order. This PR will fix one of them but I would prefer having a different fix, if possible. The limit in https://github.com/sympy/sympy/issues/13715 fails because `Order(p)` is not an instance of `Order` but 0 when `p.is_zero`. I think that changing `Order(p)` would probably be a better fix.

**Comment by sachin-4099** [comment]
I will takeover and come up with a PR soon.

**Comment by iamabhishek0** [comment]
@czgdp1807 the issue was fixed in #18629 , so this could be closed

**Comment by czgdp1807** [comment]
The issue has been fixed. 
```python
>>> n = Symbol('n')
>>> p = Symbol('p', zero=True)
>>> limit(n+p,n,0)
0
```
Closing for now.

**Pull request closed without merging**